scipy/stats/_sensitivity_analysis.py

```diff
@@ -0,0 +1,364 @@
+
+    A, B = A_B[:, :d], A_B[:, d:]
+
+    if dists is not None:
+        for d_, dist in enumerate(dists):
+            dist_rng = NumericalInversePolynomial(
+                dist, random_state=random_state
+            )
+            A[:, d_] = dist_rng.ppf(A[:, d_])
+            B[:, d_] = dist_rng.ppf(B[:, d_])
+
+    return A, B
+
+
+def sample_AB(A: np.ndarray, B: np.ndarray) -> np.ndarray:
+    """AB matrix.
+
+    AB: columns of B into A. Shape (n*d, d). e.g in 2d
+    Take A and replace 1st column with 1st column of B. You have (n, d)
+    Then A and replace 2nd column with 2nd column of B. You have (n, d)
+    Concatenate to get AB. Which means AB is (2*n, d).
+    """
+    n, d = A.shape
+    AB = np.empty((int(d*n), d))
+    for i in range(d):
+        AB[i*n:(i+1)*n, :] = np.column_stack((A[:, :i], B[:, i], A[:, i+1:]))
+
+    return AB
+
+
+def saltelli_2010(
+    f_A: np.ndarray, f_B: np.ndarray, f_AB: np.ndarray
+) -> Tuple[np.ndarray,  np.ndarray]:
+    """Saltelli2010 formulation.
+
+    .. math::
+
+        S_i = \frac{1}{N} \sum_{j=1}^N
+        f(\mathbf{B})_j (f(\mathbf{AB}^{(i)})_j - f(\mathbf{A})_j)
+
+    .. math::
+
+        S_{T_i} = \frac{1}{N} \sum_{j=1}^N
+        (f(\mathbf{A})_j - f(\mathbf{AB}^{(i)})_j)^2
+
+    References
+    ----------
+    .. [1] Saltelli, A., P. Annoni, I. Azzini, F. Campolongo, M. Ratto, and
+       S. Tarantola. "Variance based sensitivity analysis of model
+       output. Design and estimator for the total sensitivity index."
+       Computer Physics Communications, 181(2):259-270,
+       :doi:`10.1016/j.cpc.2009.09.018`, 2010.
+    """
+    f_AB = f_AB.reshape(-1, f_A.shape[0])
+
+    var = np.var(np.vstack([f_A, f_B]))
+
+    s = np.mean(f_B * (f_AB - f_A.flatten()).T, axis=0) / var
+    st = 0.5 * np.mean((f_A.flatten() - f_AB).T ** 2, axis=0) / var
+
+    return s, st
+
+
+def sobol_indices(
+    *,
+    func: Callable,
+    n: IntNumber,
+    d: IntNumber,
+    dists: Optional[List[PINVDist]] = None,
+    l_bounds: Optional[npt.ArrayLike] = None,
+    u_bounds: Optional[npt.ArrayLike] = None,
+    random_state: SeedType = None
+):
+    """Global sensitivity indices of Sobol'.
+
+    Parameters
+    ----------
+    func : callable
+        Function to compute the Sobol' indices from. It's signature must be
+        ``func(x: ArrayLike) -> ArrayLike``, with ``x`` of shape ``(-1, d)``
+        and the output should have a shape ``(-1, 1)``.
+    n : int
+        Must be a power of 2. The total number of function call will
+        ``n(d+2)``.
+    d : int
+        Dimension of the parameter space.
+    dists : list(distributions), optional
+        List of distributions of the parameters. It must be compatible with
+        `scipy.stats.sampling.NumericalInversePolynomial`.
+    l_bounds, u_bounds : array_like (d,), optional
+        Lower and upper bounds (resp. :math:`a`, :math:`b`) of transformed
+        data in the parameter space.
```

Comment:
I think this is wrong to use bounds and will lead to issues if we use something else than a uniform distribution.

It's usually really the distribution that you want to control. Not really bounds as the problem would be: param x1 is following a Normal distribution in [-1, 5].

So maybe we can force people to give `dists` only. And if we want we could cover the case of uniform dists by allowing an int. Or, other solution, we ask people to give `func(x1, x2, ...)` instead of `func(x)`. This way we can get `d`.

Reply:
I see. Yeah, I can see forcing people to use `dists`. There is a reasonable interpretation of `bounds` in the normal case though - a truncated normal distribution. But we will support truncation of all distributions in the new infrastructure, so I think we can go without `bounds` and just require `dists`.

Will people want `dists` to allow multivariate distributions? This would work if they have `rvs` methods (and we could still determine the dimensionality if we call `rvs` for all distributions in `dists`, even if some are multivariate). But I don't think we could do QMC for multivariate distributions in general.

Reply:
I went ahead and removed `d`, `bounds` so now `dists` is requested.

We don't want to allow multivariate distributions because the method needs the parameters to be independent. There are extensions that would make it work, but hard and other methods might be more appropriate instead: 'moment independent'. In real life, there are not so many people that know their distributions, and even less that know how to correctly detect/express possible dependencies between their variables. The only thing I have seen is people using Copulas.